hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -378,19 +386,53 @@ instance MonadIde IdeM where
 
   getPlugins = asks idePlugins
 
+  withProgress t f = do
+    lf <- asks ideEnvLspFuncs
+    withProgress' lf t f
+
+  withIndefiniteProgress t f = do
+    lf <- asks ideEnvLspFuncs
+    withIndefiniteProgress' lf t f
+
 instance MonadIde IdeGhcM where
   getRootPath = lift $ lift getRootPath
   getVirtualFile = lift . lift . getVirtualFile
   getConfig = lift $ lift getConfig
   getClientCapabilities = lift $ lift getClientCapabilities
   getPlugins = lift $ lift getPlugins
+  withProgress t f = do
+    lf <- lift $ lift $ asks ideEnvLspFuncs
+    withProgress' lf t f
+  withIndefiniteProgress t f = do
+    lf <- lift $ lift $ asks ideEnvLspFuncs
+    withIndefiniteProgress' lf t f
 
 instance MonadIde IdeDeferM where
   getRootPath = lift getRootPath
   getVirtualFile = lift . getVirtualFile
   getConfig = lift getConfig
   getClientCapabilities = lift getClientCapabilities
   getPlugins = lift getPlugins
+  withProgress t f = do
+    lf <- lift $ asks ideEnvLspFuncs
+    withProgress' lf t f
+  withIndefiniteProgress t f = do
+    lf <- lift $ asks ideEnvLspFuncs
+    withIndefiniteProgress' lf t f
+
+withProgress' :: MonadIO m => Maybe (Core.LspFuncs Config) -> T.Text -> ((Core.Progress -> m ()) -> m a) -> m a
+withProgress' lspFuncs t f =
+  let mWp = Core.withProgress <$> lspFuncs
```

Comment:
wouldn't it be better to pass the actual `Core.withProgress` instead of getting it here?

Reply:
And the passed in function could be of the form `a -> IO ()`, so obviating the need for `MonadIO` too.

Reply:
@Lorenzo currently there is the possibility that lspFuncs returns null, I think this is due to the existence of the JSON transport. I think this also asks the question, should we just give plugins full access to lsp funcs?

Reply:
Either give full access, or make the progress reporting function available to them by default, as in my earlier comment. I think we already do something similar with accessibility to the VFS stuff. I would have to take a look though.

I am very much in favour of constraining what plugins/monads can do, so we do not get weird stuff happening

Reply:
I've reworked the class in https://github.com/haskell/haskell-ide-engine/pull/1190/commits/8d1925dcfc8534d560fac5752ff96ac85dfce17a, so only the functions that require `MonadIO` have the constraint. @Lorenzo is it ok if once JSONStdio is gone and there is the guarantee of `LspFuncs` (#1105) I can revisit this to directly pass it instead?

Reply:
Sure, that makes sense